.github/codeql/queries/unique-command-use.ql

```diff
@@ -0,0 +1,93 @@
+
+ abstract class CommandUsage extends Locatable {
+   abstract string getCommandName();
+ 
+   predicate isUsedFromOtherPlace() {
+     exists(CommandUsage e | e != this and e.getCommandName() = this.getCommandName())
+   }
+ 
+   predicate isFirstUsage() {
+     forall(CommandUsage e | e.getCommandName() = this.getCommandName() |
+       e.getLocationOrdinal() >= this.getLocationOrdinal()
+     )
+   }
+ 
+   string getLocationOrdinal() {
+     result =
+       this.getFile().getRelativePath() + ":" + this.getLocation().getStartLine() + ":" +
+         this.getLocation().getStartColumn()
+   }
+ }
+ 
+ class CommandUsageCallExpr extends CommandUsage, CallExpr {
+   CommandUsageCallExpr() {
+     this.getCalleeName() = "executeCommand" and
+     this.getArgument(0).(StringLiteral).getValue().matches("%codeQL%") and
+     not this.getFile().getRelativePath().matches("extensions/ql-vscode/test/%")
+   }
+ 
+   override string getCommandName() { result = this.getArgument(0).(StringLiteral).getValue() }
+ }
+ 
+ class CommandUsagePackageJsonMenuItem extends CommandUsage, JsonObject {
+   CommandUsagePackageJsonMenuItem() {
+     this.getFile().getBaseName() = "package.json" and
+     exists(this.getPropValue("command")) and
+     exists(JsonObject topObject, string menuName |
+       not exists(topObject.getParent()) and
+       topObject
+           .getPropValue("contributes")
+           .getPropValue("menus")
+           .getPropValue(menuName)
+           .getElementValue(_) = this and
+       menuName != "commandPalette"
+     )
+   }
+ 
+   override string getCommandName() { result = this.getPropValue("command").getStringValue() }
+ }
+ 
+ predicate isDisabledInCommandPalette(string commandName) {
+   exists(JsonObject topObject, JsonObject commandPaletteObject |
+     not exists(topObject.getParent()) and
+     topObject
+         .getPropValue("contributes")
+         .getPropValue("menus")
+         .getPropValue("commandPalette")
+         .getElementValue(_) = commandPaletteObject and
+     commandPaletteObject.getPropValue("command").getStringValue() = commandName and
+     commandPaletteObject.getPropValue("when").getStringValue() = "false"
+   )
+ }
+ 
+ class CommandUsagePackageJsonCommandPalette extends CommandUsage, JsonObject {
+   CommandUsagePackageJsonCommandPalette() {
+     this.getFile().getBaseName() = "package.json" and
+     exists(this.getPropValue("command")) and
+     exists(JsonObject topObject |
+       not exists(topObject.getParent()) and
+       topObject.getPropValue("contributes").getPropValue("commands").getElementValue(_) = this
+     ) and
+     not isDisabledInCommandPalette(this.getPropValue("command").getStringValue())
+   }
+ 
+   override string getCommandName() { result = this.getPropValue("command").getStringValue() }
+ }
+ 
+ from CommandUsage e
+ where
+   e.isUsedFromOtherPlace() and
+   not e.isFirstUsage()
+ select e, "The " + e.getCommandName() + " command is used from another location"
```

Comment:
Skipping the first usage is clever. I can also see it being confusing if you want to audit all the places that call a command and decide which one you want to preserve (which may not be the first). As a consumer of the alert, would it be useful to just create an alert for each use? Then you don't need `isFirstUsage`. 
You can also put the total count in the message, e.g. `count(CommandUsage use | use.getCommandName() = ...)`.

Reply:
The original aim of omitting the first usage was so that going from one usage to two usages only creates one alert rather than two. But you make a very good point that the one that gets alerted isn't necessarily the one you want to remove/update. If everyone is ok with it perhaps it would be best to just alert on every instance so it's easier to find the usages.

Including the number of usages in the message is also a good idea. If you go from two usages to three, would than count as 1 alert added, or would it count as two alerts removed and three alerts added 🤔

Reply:
After discussion with @norascheuch we have decided to go the other way and instead only alert once per command. So this means if there are multiple usages we have to pick one, and we decided to stick with this existing `isFirstUsage` implementation. The exact implementation doesn't matter too much and we can change it in the future. For right now we just want this query to be useful to us and we know that the location being alerted on may not be the place to change but you should look for other usages too.

To make this easier, I've changed the alert message to include the number of usages, and added some documentation to the query indicating that you should search for other usages when fixing.